--- src/ls.c	2019-02-19 05:00:02.000000000 +0100
+++ src/ls.c	2019-07-02 14:42:03.826779900 +0200
@@ -4758,21 +4758,16 @@
         }
     }
 
-  /* Check the file's suffix only if still classified as C_FILE.  */
-  ext = NULL;
-  if (type == C_FILE)
-    {
-      /* Test if NAME has a recognized suffix.  */
+  /* Test if NAME has a recognized suffix.  */
 
-      len = strlen (name);
-      name += len;		/* Pointer to final \0.  */
-      for (ext = color_ext_list; ext != NULL; ext = ext->next)
-        {
-          if (ext->ext.len <= len
-              && c_strncasecmp (name - ext->ext.len, ext->ext.string,
-                                ext->ext.len) == 0)
-            break;
-        }
+  len = strlen (name);
+  name += len;		/* Pointer to final \0.  */
+  for (ext = color_ext_list; ext != NULL; ext = ext->next)
+    {
+      if (ext->ext.len <= len
+          && c_strncasecmp (name - ext->ext.len, ext->ext.string,
+                            ext->ext.len) == 0)
+        break;
     }
 
   /* Adjust the color for orphaned symlinks.  */
@@ -4831,6 +4826,9 @@
   if (print_scontext)
     len += 1 + (format == with_commas ? strlen (f->scontext) : scontext_width);
 
+  if (print_with_color)
+	  len += 2;
+
   len += quote_name_width (f->name, filename_quoting_options, f->quoted);
 
   if (indicator_style != none)
